src/gnatss/ops/validate.py

```diff
@@ -0,0 +1,208 @@
+from typing import Tuple
+
+import numba
+import numpy as np
+import pandas as pd
+import scipy
+from nptyping import Float64, NDArray, Shape
+from numba.typed import List as NumbaList
+
+from .. import constants
+from . import calc_std_and_verify
+from .utils import clean_zeros
+
+# Constrain matrix Q
+# Hardcoded constant for 3 transponders
+Q_MATRIX = np.array(
+    [
+        [-1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+        [0.0, -1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0],
+        [0.0, 0.0, -1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0],
+        [0.0, 0.0, 0.0, -1.0, 0.0, 0.0, 1.0, 0.0, 0.0],
+        [0.0, 0.0, 0.0, 0.0, -1.0, 0.0, 0.0, 1.0, 0.0],
+        [0.0, 0.0, 0.0, 0.0, 0.0, -1.0, 0.0, 0.0, 1.0],
+        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+    ]
+)
+
+
+@numba.njit(cache=True)
+def _calc_qmxqt(atwa: NDArray, q: NDArray) -> Tuple[NDArray, NDArray, NDArray, NDArray]:
+    mx = np.linalg.inv(atwa)
+
+    mxqt = mx @ q.T
+    qmxqt = q @ mxqt
+    qmx = q @ mx
+
+    return qmxqt, qmx, mx, mxqt
+
+
+@numba.njit(cache=True)
+def _calc_xp(x: NDArray, qmxqtiqmx: NDArray, zerr: NDArray) -> NDArray:
+    xdel = zerr.T @ qmxqtiqmx
+    xp = x + xdel
+    return xp
+
+
+@numba.njit(cache=True)
+def _calc_mxp(mx: NDArray, qmxqtiqmx: NDArray, mxqt: NDArray) -> NDArray:
+    delmx = mxqt @ qmxqtiqmx
+    mxp = mx - delmx
+    return mxp
+
+
+@numba.njit(cache=True)
+def _sum_all(input_list):
```

Comment:
What's the type of `input_list`? Is there no vectorized way of doing the sum?

Reply:
Also return type hint

Reply:
> Is there no vectorized way of doing the sum?
Using numba, I can't use `np.sum`.